File: crates/verify/ess-conformance/src/go/reading.go
// ReadingReference names an already observed event occurrence/member and a contract the adapter must validate.
type ReadingReference struct {
	Event        string `json:"event"`
	Occurrence   uint32 `json:"occurrence"`
	Field        string `json:"field"`
	DeclaredType string `json:"declared_type"`
	Contract     struct {
		Encoding string `json:"encoding"`
		Origins  []struct {
			Role   string `json:"role"`
			Offset string `json:"offset"`
		} `json:"origins"`
	} `json:"contract"`
}

// ReadingObservationRequest asks for facts, never a normalized coordinate or assertion result.
type ReadingObservationRequest struct {
	Reading     ReadingReference
	Correlation string
}

// ClockReadingTarget validates the actual declared member contract and observes its source epoch and formatter.
// Unsupported evidence must return ErrUnsupported; an authored contract is never a certificate.
type ClockReadingTarget interface {
	ObserveClockReading(ReadingObservationRequest) (ClockReadingEvidence, error)
}

func (r ReadingReference) occurrenceKey() string {
	return fmt.Sprintf("%s#%d.%s", r.Event, r.Occurrence, r.Field)
}
func admitReading(value any) error {
	f, err := closed(value, "event occurrence field declared_type contract", "")
	if err != nil {
		return err
	}
	if err = name(f["event"], false); err != nil {
		return err
	}
	if err = name(f["declared_type"], false); err != nil {
		return err
	}
	n, err := unsigned(f["occurrence"])
	if err != nil || n > 65535 {
		return fmt.Errorf("invalid reading occurrence")
	}
	field, err := text(f["field"])
	if err != nil || len(field) > 128 || !regexp.MustCompile(`^[A-Za-z][A-Za-z0-9_]*$`).MatchString(field) {
		return fmt.Errorf("invalid reading member")
	}
	event, _ := text(f["event"])
	if len(fmt.Sprintf("%s#%d.%s", event, n, field)) > 512 {
		return fmt.Errorf("reading occurrence exceeds bound")
	}
	contract, err := closed(f["contract"], "encoding origins", "")
	if err != nil {
		return err
	}
	encoding, err := text(contract["encoding"])
	if err != nil {
		return err
	}
	origins, ok := contract["origins"].([]any)
	if !ok || len(origins) == 0 || len(origins) > 3 {
		return fmt.Errorf("invalid reading origins")
	}
	previous := map[[2]string]bool{}
	for _, v := range origins {
		origin, err := closed(v, "role offset", "")
		if err != nil {
			return err
		}
		role, err := text(origin["role"])
		if err != nil {
			return err
		}
		offset, err := text(origin["offset"])
		if err != nil {
			return err
		}
		pair := [2]string{role, offset}
		if previous[pair] {
			return fmt.Errorf("duplicate reading origin")
		}
		previous[pair] = true
		if role != "producer_process" && role != "consumer_process" && role != "unknown" {
			return fmt.Errorf("invalid reading origin")
		}
		if !((encoding == "offset_date_time_text" && offset == "encoded_offset") || (encoding == "unix_seconds" && offset == "encoding_defined_epoch") || (encoding == "local_date_time_millis_literal_z" && (offset == "requires_observation" || offset == "unknown"))) {
			return fmt.Errorf("contradictory reading encoding and authority")
		}
	}
	return nil
}

func (r *run) expectReadingOrder(index int, step Step) bool {
	target, ok := r.target.(ClockReadingTarget)
	if !ok {
		r.skip("clock reading evidence unsupported")
		return false
	}
	resolve := func(reference *ReadingReference) (ClockCoordinate, error) {
		if reference == nil {
			return ClockCoordinate{}, fmt.Errorf("missing reading reference")
		}
		occurrence := uint32(0)
		var value any
		found := false
		for _, event := range r.seen {
			if event.Event == reference.Event {
				if occurrence == reference.Occurrence {
					value, found = event.Payload[reference.Field]
					break
				}
				occurrence++
			}
		}
		if !found {
			return ClockCoordinate{}, fmt.Errorf("reading event occurrence/member not observed")
		}
		scalar, ok := value.(string)
		if reference.Contract.Encoding == "unix_seconds" {
			switch v := value.(type) {
			case json.Number:
				scalar, ok = clockIntegerText(v.String())
			case float64:
				ok = !math.IsNaN(v) && !math.IsInf(v, 0) && v >= 0 && v <= 253402300799 && v == math.Trunc(v)
				if ok {
					scalar = strconv.FormatInt(int64(v), 10)
				}
			case int64:
				scalar = strconv.FormatInt(v, 10)
				ok = true
			case int:
				scalar = strconv.Itoa(v)
				ok = true
			default:
				ok = false
			}
		}
		if !ok {
			return ClockCoordinate{}, fmt.Errorf("reading scalar does not match encoding")
		}
		requestReference := *reference
		// The target owns its request, including nested origins; it cannot rewrite
		// the admitted contract used by this or a later assertion.
		requestReference.Contract.Origins = append(requestReference.Contract.Origins[:0:0], requestReference.Contract.Origins...)
		evidence, err := target.ObserveClockReading(ReadingObservationRequest{requestReference, r.correlation})
		if err != nil {
			return ClockCoordinate{}, err
		}
		origins := make([][2]string, 0, len(reference.Contract.Origins))
		for _, origin := range reference.Contract.Origins {
			origins = append(origins, [2]string{origin.Role, origin.Offset})
		}
		result, err := ResolveClockReading(reference.Contract.Encoding, origins, scalar, evidence, r.correlation, reference.occurrenceKey())
		if err != nil && strings.Contains(err.Error(), "UnknownEvidence") {
			return ClockCoordinate{}, fmt.Errorf("%w: %v", ErrUnsupported, err)
		}
		return result, err
	}
	left, err := resolve(step.ReadingLeft)
	if err == nil {
		var right ClockCoordinate
		right, err = resolve(step.ReadingRight)
		if err == nil {
			var order int
			order, err = CompareClockReadings(left, right)
			if err == nil {
				if (step.ReadingOrder == "before" && order < 0) || (step.ReadingOrder == "equal" && order == 0) || (step.ReadingOrder == "after" && order > 0) {
					return true
				}
				return r.fail(index, "clock coordinate order differs from %s", step.ReadingOrder)
			}
		}
	}
	if errors.Is(err, ErrUnsupported) || (err != nil && strings.Contains(err.Error(), "DifferentClock")) {
		r.skip("clock comparison unsupported: %v", err)
		return false
	}
	return r.fail(index, "clock reading: %v", err)
}

// Canonicalize exact mathematical integer tokens without a floating-point round trip.
func clockIntegerText(raw string) (string, bool) {
	if len(raw) > 512 || !regexp.MustCompile(`^-?(0|[1-9][0-9]*)(\.[0-9]+)?([eE][+-]?[0-9]+)?$`).MatchString(raw) {
		return "", false
	}
	negative := strings.HasPrefix(raw, "-")
	if negative {
		raw = raw[1:]
	}
	exponent := 0
	if index := strings.IndexAny(raw, "eE"); index >= 0 {
		var err error
		exponent, err = strconv.Atoi(raw[index+1:])
		if err != nil || exponent < -512 || exponent > 512 {
			return "", false
		}
		raw = raw[:index]
	}
	if index := strings.IndexByte(raw, '.'); index >= 0 {
		exponent -= len(raw) - index - 1
		raw = raw[:index] + raw[index+1:]
	}
	raw = strings.TrimLeft(raw, "0")
	if raw == "" {
		return "0", true
	}
	for exponent < 0 && strings.HasSuffix(raw, "0") {
		raw = strings.TrimSuffix(raw, "0")
		exponent++
	}
	if exponent < 0 || negative || len(raw)+exponent > 12 {
		return "", false
	}
	return raw + strings.Repeat("0", exponent), true
}

File: crates/verify/ess-conformance/src/ts/reading.ts
// Reading conformance: an already observed event member, and separately observed clock facts.
//
// Ported from `src/go/reading.go`. The accessor never reads a clock and never normalizes anything
// itself — it hands the adapter a request for *facts* and gives those facts, with the contract the
// suite admitted, to the coordinate reader. An authored contract is a claim, never a certificate,
// which is why the origins handed to the target are a copy: a target that rewrote them would be
// deciding the terms of its own assertion.

import {
  ErrUnsupported,
  JsonNumber,
  closed,
  errorText,
  isClockReadingTarget,
  isUnsupported,
  name,
  text,
  unsigned,
  unsupported,
} from './runtime.js';
import type { Node, ScenarioRun, Step } from './runtime.js';
import {
  compareClockReadings,
  resolveClockReading,
  type ClockCoordinate,
  type ClockReadingEvidence,
  type ReadingOriginPair,
} from './coordinate.js';

/** One declared `(role, offset-authority)` alternative of a reading contract. */
export interface ReadingOrigin {
  role: string;
  offset: string;
}

/** The encoding a member is written in, and the authorities that may have produced it. */
export interface ReadingContract {
  encoding: string;
  origins: ReadingOrigin[];
}

/**
 * Names an already observed event occurrence and member, and the contract the adapter must
 * validate against.
 */
export interface ReadingReference {
  event: string;
  occurrence: number;
  field: string;
  /**
   * The member's declared type, under the name the suite document writes it with.
   *
   * Snake case because this *is* the suite document: `decodeStep` casts the parsed JSON, which is
   * the plain `json.Unmarshal` the Go runtime does at the same point, and a camel-cased field
   * would silently be `undefined` in every request a target is handed.
   */
  declared_type: string;
  contract: ReadingContract;
}

/** Asks for facts, never for a normalized coordinate and never for an assertion result. */
export interface ReadingObservationRequest {
  reading: ReadingReference;
  correlation: string;
}

/**
 * What a target offers so a scenario can order two clock readings.
 *
 * Optional, like `Clock` and `OrderedReader`: a target that does not implement it is *skipped* on
 * the scenarios that need it, never failed. Evidence it cannot observe is answered by raising an
 * error carrying `ErrUnsupported`; an authored contract is never a certificate.
 */
export interface ClockReadingTarget {
  observeClockReading(request: ReadingObservationRequest): ClockReadingEvidence;
}

const READING_MEMBER = /^[A-Za-z][A-Za-z0-9_]*$/;
const CLOCK_INTEGER = /^-?(0|[1-9][0-9]*)(\.[0-9]+)?([eE][+-]?[0-9]+)?$/;

/** The key one occurrence of one member is known by, in every message and every request. */
export function occurrenceKey(reference: ReadingReference): string {
  return `${reference.event}#${reference.occurrence}.${reference.field}`;
}

/**
 * Admit one authored reading reference.
 *
 * The encoding and the declared authority have to agree: text that carries its own offset cannot
 * be produced by an epoch-defined formatter, and a literal-Z local time cannot carry one at all.
 * A contract that says otherwise describes no implementation and is refused here rather than
 * producing a verdict later.
 */
export function admitReading(value: unknown): void {
  const fields = closed(value, 'event occurrence field declared_type contract', '');
  name(fields.event, false);
  name(fields.declared_type, false);
  let occurrence: bigint | number;
  try {
    occurrence = unsigned(fields.occurrence);
  } catch {
    throw new Error('invalid reading occurrence');
  }
  if (Number(occurrence) > 65535) {
    throw new Error('invalid reading occurrence');
  }
  let field: string;
  try {
    field = text(fields.field);
  } catch {
    throw new Error('invalid reading member');
  }
  if (field.length > 128 || !READING_MEMBER.test(field)) {
    throw new Error('invalid reading member');
  }
  const event = typeof fields.event === 'string' ? fields.event : '';
  if (`${event}#${occurrence}.${field}`.length > 512) {
    throw new Error('reading occurrence exceeds bound');
  }
  const contract = closed(fields.contract, 'encoding origins', '');
  const encoding = text(contract.encoding);
  const origins = contract.origins;
  if (!Array.isArray(origins) || origins.length === 0 || origins.length > 3) {
    throw new Error('invalid reading origins');
  }
  const previous = new Set<string>();
  for (const value of origins) {
    const origin = closed(value, 'role offset', '');
    const role = text(origin.role);
    const offset = text(origin.offset);
    const pair = JSON.stringify([role, offset]);
    if (previous.has(pair)) {
      throw new Error('duplicate reading origin');
    }
    previous.add(pair);
    if (role !== 'producer_process' && role !== 'consumer_process' && role !== 'unknown') {
      throw new Error('invalid reading origin');
    }
    if (
      !(
        (encoding === 'offset_date_time_text' && offset === 'encoded_offset') ||
        (encoding === 'unix_seconds' && offset === 'encoding_defined_epoch') ||
        (encoding === 'local_date_time_millis_literal_z' &&
          (offset === 'requires_observation' || offset === 'unknown'))
      )
    ) {
      throw new Error('contradictory reading encoding and authority');
    }
  }
}

/**
 * What `expectReadingOrder` reads of the run it is handed.
 *
 * Named from unit A's own `ScenarioRun` rather than restated, so a member that moves there is a
 * type error here instead of an `undefined` at the first reading. `skip` is widened to `void`
 * because a stand-in in a case does not have to abort the way the real one does.
 */
export type ReadingRun = Pick<ScenarioRun, 'target' | 'correlation' | 'seen' | 'fail'> & {
  skip(message: string): void;
};

/** Order two observed readings, or say why the claim cannot be decided. */
export function expectReadingOrder(run: ReadingRun, index: number, step: Step): boolean {
  // Go asks with a type assertion; TypeScript has no interface at run time, so the capability is
  // the presence of the method — and `isClockReadingTarget` is the one place that is decided, for
  // this capability and for the three beside it. A second probe here could drift from it.
  if (!isClockReadingTarget(run.target)) {
    run.skip('clock reading evidence unsupported');
    return false;
  }
  const target = run.target as unknown as ClockReadingTarget;
  const resolve = (reference: ReadingReference | null | undefined): ClockCoordinate => {
    if (reference === null || reference === undefined) {
      throw new Error('missing reading reference');
    }
    let occurrence = 0;
    let value: Node;
    let found = false;
    for (const event of run.seen) {
      if (event.event === reference.event) {
        if (occurrence === reference.occurrence) {
          found = Object.hasOwn(event.payload, reference.field);
          value = found ? event.payload[reference.field] : undefined;
          break;
        }
        occurrence += 1;
      }
    }
    if (!found) {
      throw new Error('reading event occurrence/member not observed');
    }
    let scalar = typeof value === 'string' ? value : '';
    let admitted = typeof value === 'string';
    if (reference.contract.encoding === 'unix_seconds') {
      if (value instanceof JsonNumber) {
        const canonical = clockIntegerText(value.toString());
        admitted = canonical !== null;
        scalar = canonical ?? '';
      } else if (typeof value === 'number') {
        // Go's float64 arm: a whole number inside the admitted range, never rounded into one.
        admitted =
          Number.isFinite(value) &&
          value >= 0 &&
          value <= 253402300799 &&
          value === Math.trunc(value);
        scalar = admitted ? String(value) : '';
      } else if (typeof value === 'bigint') {
        // Go's int64 arm, which states the value and leaves the range to the coordinate reader.
        scalar = value.toString();
        admitted = true;
      } else {
        admitted = false;
      }
    }
    if (!admitted) {
      throw new Error('reading scalar does not match encoding');
    }
    // The target owns its request, including the nested origins; it cannot rewrite the admitted
    // contract this or a later assertion is decided against.
    const requested: ReadingReference = {
      ...reference,
      contract: {
        encoding: reference.contract.encoding,
        origins: reference.contract.origins.map((origin) => ({ ...origin })),
      },
    };
    const evidence = target.observeClockReading({
      reading: requested,
      correlation: run.correlation,
    });
    const origins: ReadingOriginPair[] = reference.contract.origins.map((origin) => [
      origin.role,
      origin.offset,
    ]);
    try {
      return resolveClockReading(
        reference.contract.encoding,
        origins,
        scalar,
        evidence,
        run.correlation,
        occurrenceKey(reference),
      );
    } catch (error) {
      if (errorText(error).includes('UnknownEvidence')) {
        // `fmt.Errorf("%w: %v", ErrUnsupported, err)`: the sentinel's own sentence, then the cause.
        throw unsupported(`${ErrUnsupported.message}: ${errorText(error)}`);
      }
      throw error;
    }
  };

  let failure: unknown;
  try {
    const left = resolve(step.left);
    const right = resolve(step.right);
    const order = compareClockReadings(left, right);
    if (
      (step.order === 'before' && order < 0) ||
      (step.order === 'equal' && order === 0) ||
      (step.order === 'after' && order > 0)
    ) {
      return true;
    }
    return run.fail(index, `clock coordinate order differs from ${step.order}`);
  } catch (error) {
    failure = error;
  }
  if (isUnsupported(failure) || errorText(failure).includes('DifferentClock')) {
    run.skip(`clock comparison unsupported: ${errorText(failure)}`);
    return false;
  }
  return run.fail(index, `clock reading: ${errorText(failure)}`);
}

/**
 * Canonicalize an exact mathematical integer token without a floating-point round trip.
 *
 * A reading that arrives as `1.709296245e9` is the same second as `1709296245`, and one that
 * arrives as `1.5` is not a second at all. Answering that by parsing to a binary64 would admit a
 * value the specification says is not an integer, so the digits are moved rather than converted.
 */
export function clockIntegerText(raw: string): string | null {
  if (raw.length > 512 || !CLOCK_INTEGER.test(raw)) {
    return null;
  }
  const negative = raw.startsWith('-');
  let digits = negative ? raw.slice(1) : raw;
  let exponent = 0;
  const marker = digits.search(/[eE]/);
  if (marker >= 0) {
    const parsed = Number(digits.slice(marker + 1));
    if (!Number.isInteger(parsed) || parsed < -512 || parsed > 512) {
      return null;
    }
    exponent = parsed;
    digits = digits.slice(0, marker);
  }
  const point = digits.indexOf('.');
  if (point >= 0) {
    exponent -= digits.length - point - 1;
    digits = digits.slice(0, point) + digits.slice(point + 1);
  }
  digits = digits.replace(/^0+/, '');
  if (digits === '') {
    return '0';
  }
  while (exponent < 0 && digits.endsWith('0')) {
    digits = digits.slice(0, -1);
    exponent += 1;
  }
  if (exponent < 0 || negative || digits.length + exponent > 12) {
    return null;
  }
  return digits + '0'.repeat(exponent);
}
